SVPAlib/src/automata/sra/SRA.java

```diff
@@ -0,0 +1,1828 @@
+/**
+ * SVPAlib
+ * automata.sra
+ * Jul 25, 2018
+ * @author Tiago Ferreira
+ */
+package automata.sra;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.Stack;
+
+import org.sat4j.specs.TimeoutException;
+
+import automata.Automaton;
+import automata.Move;
+import theory.BooleanAlgebra;
+import utilities.Block;
+import utilities.Pair;
+import utilities.Timers;
+import utilities.UnionFindHopKarp;
+
+/**
+ * Symbolic Register Automaton
+ * 
+ * @param <P> set of predicates over the domain S
+ * @param <S> domain of the automaton alphabet
+ */
+public class SRA<P, S> extends Automaton<P, S> {
+	// ------------------------------------------------------
+	// Constant automata
+	// ------------------------------------------------------
+
+	public void setIsDet(boolean b) {
+		isDeterministic = b;
+	}
+
+	/**
+	 * Returns the empty SRA for the Boolean algebra <code>ba</code>
+	 * @throws TimeoutException 
+	 */
+	public static <A, B> SRA<A, B> getEmptySRA(BooleanAlgebra<A, B> ba) throws TimeoutException {
+		SRA<A, B> aut = new SRA<A, B>();
+		aut.states = new HashSet<Integer>();
+		aut.states.add(0);
+		aut.finalStates = new HashSet<Integer>();
+		aut.initialState = 0;
+        aut.registers = new HashSet<A>();
+        aut.registers.add(ba.False());
+		aut.isDeterministic = true;
+		aut.isEmpty = true;
+		aut.maxStateId = 1;
+        // FIXME: We require a register operation on every transition.
```

Comment:
For an empty SRA I added the value of the register as `ba.False()`, so that our input is never in the register.

Reply:
Why does an empty SRA have registers?

Reply:
I just made the assumption that SRAs have at least one register. As every move mandatorily has a register operation, it must affect at least one register.

Reply:
Don't you the special "doesn't match any register" move? If you have that, in the absence of registers that's the move to take

Reply:
That’s the fresh operation, if it doesn’t match any register, we write it to the register passed, so we’d still need the register